### PR TITLE
Use ip

### PR DESCRIPTION
**Pull request by kyle-long**
The command ip

### CODE REVIEW

File: formulas/hostname/files/set-hostname [hunk 1]
```diff
@@ -19,6 +19,8 @@
 . /usr/local/lib/wick-infect
 wickStrictMode
 
+declare ip iface help verbose full reloadFlag
```
Comment (fidian) [inline formulas/hostname/files/set-hostname]: Would you alphabetize these please?

File: lib/wick-get-iface-ip [hunk 2]
```diff
@@ -20,36 +20,58 @@
 #
 # Returns 0 on success, non-zero on failure.
 wickGetIfaceIp() {
-    local ip list
+    local list target interface allInterfaces
```
Comment (fidian) [inline lib/wick-get-iface-ip]: Another place to alphabetize.

File: lib/wick-get-iface-ip [hunk 3]
```diff
@@ -20,36 +20,58 @@
...
+    fi
+
+    if wickCommandExists ip; then
+        cmd=(ip addr show)
+
+        if [[ -n "$interface" ]]; then
+            cmd[${#cmd[@]}]="$interface"
+        fi
+
+        list=( $("${cmd[@]}" 2>/dev/null | grep -Po "inet\s+\K[\d.]+") )
```
Comment (fidian) [inline lib/wick-get-iface-ip]: I am concerned about `grep -Po` working on a Mac.  Have you tested this?

File: lib/wick-get-iface-ip [hunk 4]
```diff
@@ -20,36 +20,58 @@
...
+    else
+        wickError "Couldn't find a command to look up ip addresses"
+
+        return 2
+    fi
+
+    if [[ "$allInterfaces" == true ]]; then
         if [[ ${#list[@]} -eq 0 ]]; then
-            local "$1" && wickIndirectArray "$1"
+            local "$target" && wickIndirectArray "$target"
```
Comment (fidian) [inline lib/wick-get-iface-ip]: This is a discussion point only.  It is inspired by the code review but would certainly not block this code from being accepted.

What do you think about this syntax?  I find it bulky.  I have a suggestion and I'm willing to implement it, but it certainly needs some thought and review by people (like you!).  Here's the suggestion:

* Remove the "if [[ ${#list[@]} -eq 0 ]]" block entirely.
* Change wickIndirectArray to contain that block.
* Add another parameter to wickIndirectArray so we pass the number of elements.

    local "$dest" && wickIndirectArray "$dest" ${#list[@]} "${list[@]}"

Perhaps we could do even more magic?  We could just pass the variable name and the scope of wickIndirectArray could, in theory, just grab from that variable.  We would need to remember to not use `$` in front.

    wickIndirectArray "$destinationVariable" list

Thoughts?  Alternatives?

File: lib/wick-get-iface-ip [hunk 5]
```diff
@@ -20,36 +20,58 @@
...
         local "$1" && wickIndirectArray "$1" "${list[@]}"
+
         return 0
     fi
 
-    ip=$(ifconfig -v $2 2>/dev/null | grep inet\ | head -n 1 || echo "")
-    ip=${ip#*inet addr:}
-    ip=${ip#*inet }
-    ip=${ip%% *}
+    if [[ ${list[@]} -lt 1 ]]; then
```
Comment (fidian) [inline lib/wick-get-iface-ip]: This looks wrong.  You probably need a `#` in there.

File: tests/lib/wick-get-iface-ip.bats [hunk 6]
```diff
@@ -86,5 +104,50 @@ setup() {
...
 
     mock-command ifconfig wick-get-iface-ip/bsd
+    mock-command wickCommandExists wick-get-iface-ip/wick-command-exists-no-ip
     ! wickGetIfaceIp RESULT noip
 }
+
+@test "lib/wick-get-iface-ip: ip command - All interfaces" {
+    local result
+
+    mock-command ip wick-get-iface-ip/ip
```
Comment (fidian) [inline tests/lib/wick-get-iface-ip.bats]: Is the `ip` commant not available on a mac?  Or does it have different output than Gnu/Linux?

**Comment by fidian** [comment]
Reviewed and there are a couple concerns.

File: formulas/hostname/files/set-hostname (hunk 1 above)
Comment (kyle-long) [inline formulas/hostname/files/set-hostname]: c689ee142e4b1f3b129321e91742d9b7622cc0d4

File: lib/wick-get-iface-ip (hunk 2 above)
Comment (kyle-long) [inline lib/wick-get-iface-ip]: d1e9d6bb61c3dfb233ffb779de80f66bdb4b492a

File: lib/wick-get-iface-ip (hunk 3 above)
Comment (kyle-long) [inline lib/wick-get-iface-ip]: Good catch.  It -P doesn't exist for bsd grep

5fce15c99f1d87bbb8177a92f6ed115f9cf4666d

File: lib/wick-get-iface-ip (hunk 5 above)
Comment (kyle-long) [inline lib/wick-get-iface-ip]: Yes my mistake

72e7831b15c607fc77cb8c7789778b6ac9fc84b9

File: lib/wick-get-iface-ip (hunk 4 above)
Comment (kyle-long) [inline lib/wick-get-iface-ip]: Could you do magic like check the number of arguments passed to wickIndirectArray and check if those arguments are empty?  Or is it a use case to have an array like

```
myArray=(
    ""
    hello
    ""
)
```

File: lib/wick-get-iface-ip [hunk 7]
```diff
@@ -20,36 +20,58 @@
...
+
+    if [[ "$allInterfaces" == true ]]; then
         if [[ ${#list[@]} -eq 0 ]]; then
-            local "$1" && wickIndirectArray "$1"
+            local "$target" && wickIndirectArray "$target"
+
             return 0
         fi
 
         local "$1" && wickIndirectArray "$1" "${list[@]}"
```
Comment (fidian) [inline lib/wick-get-iface-ip]: This should use `$target` not `$1`.

File: lib/wick-get-iface-ip (hunk 4 above)
Comment (fidian) [inline lib/wick-get-iface-ip]: I currently do the magic-like check.  What I would like to do is this:

    myArray=()
    local "$dest" && wickIndirectArray "$dest" "${myArray[@]-?}"

However this sends an empty string if the array is empty.

    local "$dest" && wickIndirectByRef "$dest" "myArray"

Well, I will document this and put it in a ticket.  Consider this resolved.

File: lib/wick-get-iface-ip (hunk 7 above)
Comment (kyle-long) [inline lib/wick-get-iface-ip]: 627d5f8ae87473b63b0174a48c034445dc617b77

**Comment by fidian** [comment]
Yes, works for me.  Squashing into one beautiful commit.

**Pull request merged**